src/fem/discrete_space.hh

```diff
@@ -0,0 +1,110 @@
+/*!
+ *
+﻿ * Copyright (C) 2015 Technical University of Liberec.  All rights reserved.
+ * 
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 3 as published by the
+ * Free Software Foundation. (http://www.gnu.org/licenses/gpl-3.0.en.html)
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ *
+ * 
+ * @file    dofhandler.hh
+ * @brief   Declaration of class which handles the ordering of degrees of freedom (dof) and mappings between local and global dofs.
+ * @author  Jan Stebel
+ */
+
+#ifndef DISCRETE_SPACE_HH_
+#define DISCRETE_SPACE_HH_
+
+#include <map>
+#include <petscmat.h>
+#include "mesh/mesh_types.hh"
+#include "mesh/elements.h"
+#include "la/distribution.hh"
+#include "fem/finite_element.hh"
+
+
+
+template<unsigned int dim, unsigned int spacedim> class FiniteElement;
+class Mesh;
+class Distribution;
+
+
+/**
+ * Abstract class for definition of finite element functions on the mesh.
+ * This should include
+ * - simple FE spaces using the same finite element on
+ *   all mesh elements with the same dimension,
+ * - p-refined spaces using variable FE order,
+ * - XFEM with arbitrary shape functions on every element.
+ * 
+ */
+class DiscreteSpace {
+public:
+    
+  /// Number of dofs associated to node. @p nid is the node index in the mesh tree.
+  virtual unsigned int n_node_dofs(unsigned int nid) const = 0;
+  
```

Comment:
Methods for dofs on edges and faces are not implemented yet (depends also on DuplicateNodes).

The number of dofs for one n-face (point/line/triangle) must be unique (elements sharing this n-face may not have different number of dofs on it).